### ISSUE TITLE
Marketing Upsell: can't decline offer from Editor Checkout

### ISSUE THREAD

**Issue by michaeldcain** [Automattic/wp-calypso#50461]
This was partially hidden by #50450, which is being fixed in #50460. You'll need to work from the latter to reproduce this until it's merged.

### Steps to reproduce the behavior

1. On a free site, go to the editor
2. Add a premium block (i.e. Pay with PayPal)
3. Click the upgrade nudge
4. Complete Checkout for the Premium plan
5. You'll be shown the Marketing upsell
6. Try to decline the upsell
7. Get frustrated

#### What I expected to happen

The offer is declined and I am returned to the editor (using the `redirect_to` query param).

#### What actually happened

Nothing.

#### Level of impact (Does it block purchases? Does it affect more than just one site?)

This essentially blocks users from returning to what they were doing without purchases a much more expensive plan.

#### Screenshot / Video: If applicable, add screenshots to help explain your problem.

https://user-images.githubusercontent.com/942359/109072987-c8882d00-76c3-11eb-9134-ae6b15f89985.mov

@Automattic/martech @Automattic/martech-decepticons @Automattic/shilling  @razvanpapadopol

**Comment by ramonjd** [comment]
@niranjan-uma-shankar I think I need some help here as I'm terribly out of context. 

When testing I see that the URL returned is from `getUrlWithQueryParam` added in this [PR](https://github.com/Automattic/wp-calypso/commit/768e81f9cbd8aec3d7be28d95c9cd728391c716d#diff-ea2f96cfb7728d99fffeb9644dbda281c9225548684c9acf8b918e396ef0e5dbR209)

Debug output for getThankYouPageUrl:

```
calypso:composite-checkout:get-thank-you-page-url starting getThankYouPageUrl +17s
common.js:109 calypso:composite-checkout:get-thank-you-page-url pendingOrReceiptId is +0ms 5444xxxx
common.js:109 calypso:composite-checkout:get-thank-you-page-url site with receipt or cart; feature is +1ms undefined
common.js:109 calypso:composite-checkout:get-thank-you-page-url fallbackUrl is +0ms /checkout/thank-you/MY_SITE/54445115
common.js:109 calypso:composite-checkout:get-thank-you-page-url cookie url is +1ms http://calypso.localhost:3000/checkout/MY_SITE#step2
calypso:composite-checkout:get-thank-you-page-url is eligible for signup destination +0ms http://calypso.localhost:3000/checkout/MY_SITE#step2
```
getThankYouPageUrl is returning the same url as we're on so `page.redirect()` doesn't take us anywhere  🤷 

I've been testing over at https://github.com/Automattic/wp-calypso/pull/50469 but haven't come up with a definitive solution yet.

cc @sirbrillig, whose name I also plucked from the annotations! :D

**Comment by ramonjd** [comment]
I wonder if it's because we're not clearing the [destination cookie](https://github.com/Automattic/wp-calypso/blob/a5683f943015772972ea8df93462c559797a548a/client/my-sites/checkout/upsell-nudge/index.jsx#L276)?

When `isEligibleForSignupDestinationResult` [returns the cookie value](https://github.com/Automattic/wp-calypso/blob/0378d16ae2befc282080b7b9c3fc61b3f4c75ebf/client/state/selectors/is-eligible-for-signup-destination.js),  it tells [this function](https://github.com/Automattic/wp-calypso/blob/768e81f9cbd8aec3d7be28d95c9cd728391c716d/client/my-sites/checkout/composite-checkout/hooks/use-get-thank-you-url/get-thank-you-page-url.ts#L205) to return the same url with query parameter.

**Comment by sirbrillig** [comment]
It sounds like you're on the right track. Despite my name being in there, I don't actually know much about the intended logic of the thank-you url function. It's pretty much all @Automattic/martech About a year ago we found that the function was often giving unexpected results in the old version of checkout (like this case!) and to help answer questions about why, I moved it to its own file and covered its output with unit tests and debug statements. 

The fact that cookies can override its output is actually really problematic because it means that no amount of tests will be able to cover its possible results and because the cookie is not keyed by site or user so it could possibly affect other sites. I actually wrote about this problem in an internal p2 at pbOQVh-lb-p2 suggesting that we get rid of the cookie entirely although we never got very far apart from removing the cookie in more places. Primarily the issue is that signup and Jetpack both use the cookie in ways that are not well understood and despite being very far away from checkout they need a way to override the thank-you page.

**Comment by niranjan-uma-shankar** [comment]
Here are my findings thus far:

1. The issue is stemming from the following line which sets a special redirect URL if the user is making an "in editor" purchase. It appears the intent is to get the editor page URL of the form `/page/{_site_slug_}/{page_id}` but  instead `window?.location.href` is taking the value `/checkout/{_site_slug_}#step2`. When the decline button is clicked on the upsell offer page, it then attempts to redirect to this checkout page which is failing.

https://github.com/Automattic/wp-calypso/blob/fc3f8f6a07df68b24b0c1d553520586c96f73bf5/client/my-sites/checkout/composite-checkout/hooks/use-get-thank-you-url/get-thank-you-page-url.ts#L145

2. I suspect that the `window?.location.href` value is getting affected by the checkout overlay feature which is not returning the page editor URL.

3. To confirm (2), you can try the following steps, Looks like a bug:
    - Click on "Upgrade" from a premium block, wait for the checkout overlay to load
    - Click the close button to exit checkout
    - You're now in the editor(notice the browser URL is not the editor URL though). 
    - Reload the page and notice that you will be taken to the standalone checkout page, while you would expect to still be in the editor. 
    
    
![checkoutoverlay](https://user-images.githubusercontent.com/1269602/109276961-ad8cf880-783c-11eb-85fa-2ace7316c556.gif)

    
    
    
4. I tried replacing [this line](https://github.com/Automattic/wp-calypso/blob/trunk/client/my-sites/checkout/composite-checkout/hooks/use-get-thank-you-url/get-thank-you-page-url.ts#L145) with the `previousRoute` value:
```
saveUrlToCookie( previousRoute );
```
However, this takes me to `/pages/{site_slug}` which is one level before the editor. 

I haven't dug into the checkout overlay code but does it affect the history state in some way? Is that why the `previousRoute` value is not picking the page editor URL(`/page/site_slug/pageId`)? More insights here would help and maybe fix the issue seen in point (3). If that is fixed, then the change proposed in point(4) might fix the bug reported in this issue. I'm wondering if any other change to the [getThankYouPageUrl()](https://github.com/Automattic/wp-calypso/blob/trunk/client/my-sites/checkout/composite-checkout/hooks/use-get-thank-you-url/get-thank-you-page-url.ts#L45) function would end up being a band-aid if we don't address the underlying cause.

cc'ing @yansern @alshakero on the checkout overlay issue in point(3) (please direct me if someone else is more appropriate, I picked from the GH annotations).

**Comment by yansern** [comment]
@niranjan-uma-shankar I really appreciate the detailed digging. I'm going to start by fixing point 3 - where hitting Close button doesn't revert to the original route, and then see if that fixes point 4.

**Comment by razvanpapadopol** [comment]
> @niranjan-uma-shankar I really appreciate the detailed digging. I'm going to start by fixing point 3 - where hitting Close button doesn't revert to the original route, and then see if that fixes point 4.

I believe point 3 was introduced by this change: https://github.com/Automattic/wp-calypso/pull/49440/files#diff-e02993c83d57a8f4e7f1626e2fd7bf965a47e26ac45cab6c2dd9bac2a49abc26

**Comment by yansern** [comment]
This PR fixes point 3:
- https://github.com/Automattic/wp-calypso/pull/50521

This PR fixes Decline button not working (would recommend waiting for https://github.com/Automattic/wp-calypso/pull/50521 to be merged).
- https://github.com/Automattic/wp-calypso/pull/50524

**Comment by ciampo** [comment]
@razvanpapadopol @yansern Can this issue be considered fixed? It looks like both #50521 and #50524 have been merged

**Comment by yansern** [comment]
Yes with both merged this is considered fixed. Closing issue!
